cmd/exo/cmd/firewall_add.go

```diff
@@ -42,49 +65,183 @@ func firewallAddRun(cmd *cobra.Command, args []string) {
+	}
+
+	sg, err := cmd.Flags().GetString("security-group")
+	if err != nil {
+		log.Fatal(err)
 	}
 
-	securGrp, err := getSecuGrpWithNameOrID(cs, sg)
+	sgs := getCommaflag(sg)
+
+	userSecurityGroups, err := getSGs(cs, sgs)
 	if err != nil {
 		log.Fatal(err)
 	}
 
-	if cidr != "" {
-		rule.Cidr = cidr
+	rule.UserSecurityGroupList = userSecurityGroups
+
+	icmptype, err := getUint8CustomFlag(cmd, "icmp-type")
+	if err != nil {
+		log.Fatal(err)
+	}
+	icmpcode, err := getUint8CustomFlag(cmd, "icmp-code")
+	if err != nil {
+		log.Fatal(err)
 	}
 
-	req := &egoscale.AuthorizeSecurityGroupIngress{
-		Protocol:        rule.Protocol,
-		CidrList:        []string{rule.Cidr},
-		SecurityGroupID: securGrp.ID,
-		Description:     rule.Description,
+	if icmptype.uint8 != nil {
+		rule.IcmpType = *icmptype.uint8
 	}
 
-	if rule.Protocol == "icmp" || rule.Protocol == "icmpv6" {
-		req.IcmpType = rule.IcmpType
-		req.IcmpCode = rule.IcmpCode
-	} else {
-		req.StartPort = rule.StartPort
-		req.EndPort = rule.EndPort
+	if icmpcode.uint8 != nil {
+		rule.IcmpCode = *icmpcode.uint8
 	}
 
-	_, err = cs.Request(req)
+	isEgress, err := cmd.Flags().GetBool("engress")
 	if err != nil {
 		log.Fatal(err)
 	}
-	firewallDetails(sg)
+
+	port, err := cmd.Flags().GetString("port")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ports := getCommaflag(port)
+
+	portsRange, err := getPortsRange(ports)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, portRange := range portsRange {
+
+		rule.StartPort = portRange.start
+		rule.EndPort = portRange.end
+		_, err = addRule(rule, isEgress)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	firewallDetails(rule.SecurityGroupID)
+}
+
+type portRange struct {
+	start uint16
+	end   uint16
+}
+
+func getPortsRange(ports []string) ([]portRange, error) {
+	portsRange := make([]portRange, len(ports))
+	for i, p := range ports {
+		pRange := strings.Split(p, "-")
+		if len(pRange) > 2 || len(pRange) == 0 {
+			return nil, fmt.Errorf("failed to get ports")
+		}
+		p1, err := strconv.ParseUint(pRange[0], 10, 16)
+		if err != nil {
+			return nil, err
+		}
+		portsRange[i].start = uint16(p1)
+		portsRange[i].end = uint16(p1)
+		if len(pRange) == 2 {
+			p2, err := strconv.ParseUint(pRange[0], 10, 16)
+			if err != nil {
+				return nil, err
+			}
+			portsRange[i].end = uint16(p2)
+		}
+	}
+	return portsRange, nil
+}
+
+func getSGs(cs *egoscale.Client, sgs []string) ([]egoscale.UserSecurityGroup, error) {
+
+	sgRes := []egoscale.UserSecurityGroup{}
+
+	for _, sg := range sgs {
+
+		s, err := getSecuGrpWithNameOrID(cs, sg)
+		if err != nil {
+			return nil, err
+		}
+		sgRes = append(sgRes, egoscale.UserSecurityGroup{Group: s.Name, Account: s.Account})
+	}
+	return sgRes, nil
+}
+
+func getDefaultRule(ruleName string) (*egoscale.AuthorizeSecurityGroupIngress, error) {
+	ruleName = strings.ToLower(ruleName)
+	if ruleName == "ping" {
+		return &egoscale.AuthorizeSecurityGroupIngress{
+			Protocol:    "icmp",
+			CidrList:    []string{"0.0.0.0/0"},
```

Comment:
I'd create some constants for the open IPv4 and IPv6 CIDR values.

Reply:
👍